### PR TITLE
Implement application name getting and setting

### LINKED ISSUE #81: Allow specifying application name in PulseAudio context creation
Hi, I’m using the audio visualizer [panon](https://github.com/rbn42/panon#readme). Panon uses this project to grab audio data from a monitor of the current output. 

My desktop, KDE Plasma, shows a little microphone symbol when something is using a microphone or monitor. Hovering it while panon is running says “audio is using the microphone”*, with the “audio” part being pretty confusing, and possibly scary for people using panon.

I grepped your source code to find out that you hardcode this string in your `_PulseAudio` backend class:

https://github.com/bastibe/SoundCard/blob/08cb467a6421c2cdf273fe55c4f1f5b9303d8d5b/soundcard/pulseaudio.py#L66

Could you please make it configurable so applications using this project can tell PulseAudio their name? Optimally this would mean that multiple contexts can be created, but using `pa_context_set_name` for the global context is probably sufficient for most applications.

*The fact that using a microphone symbol and message is wrong when using a monitor is besides the point here, but tracked in [KDE bug 410637](https://bugs.kde.org/show_bug.cgi?id=410637) if anyone is interested

### PR DESCRIPTION
**Pull request by flying-sheep**
Fixes #81

Fun fact: This is the 3rd time in my life that a non-trivial piece of code in unfamiliar territory worked on the first try.

### CODE REVIEW

File: soundcard/pulseaudio.py [hunk 1]
```diff
@@ -86,6 +86,33 @@ def _block_operation(self, operation):
         while self._pa_operation_get_state(operation) == _pa.PA_OPERATION_RUNNING:
             time.sleep(0.001)
 
+    @property
+    def name(self):
+        """Return application name stored in client proplist"""
+        idx = self._pa_context_get_index(self.context)
+        # TODO: idx can be PA_INVALID_INDEX now
```
Comment (flying-sheep) [inline soundcard/pulseaudio.py:93]: How are constants like this accessible from Python if at all?

I think we should raise an exception here if it fails to retrieve the index.

File: soundcard/pulseaudio.py [hunk 2]
```diff
@@ -86,6 +86,33 @@ def _block_operation(self, operation):
...
+        name = None
+        @_ffi.callback("pa_client_info_cb_t")
+        def callback(context, client_info, eol, userdata):
+            nonlocal name
+            if not eol:
+                name = _ffi.string(client_info.name).decode('utf-8')
+        self._pa_context_get_client_info(self.context, idx, callback, _ffi.NULL)
+        assert name is not None
```
Comment (flying-sheep) [inline soundcard/pulseaudio.py:101]: What should happen if `eol` is truthy? Exception? Or is this a programming error and an AssertionError is acceptable?

File: soundcard/pulseaudio.py (hunk 1 above)
Comment (bastibe) [reply soundcard/pulseaudio.py:93]: Usually, such constants are defined in enums, which are accessible as `_pa.CONSTANT_NAME` (look at `_pa.PA_SAMPLE_FLOAT32LE` for an example).

File: soundcard/pulseaudio.py (hunk 2 above)
Comment (bastibe) [reply soundcard/pulseaudio.py:101]: That depends on pulse's API, I would say.

Does pulse guarantee/enforce a valid UTF-8 name? If so, the assertion is fine by me. If not, returning an empty string is probably preferable. (Can you set the name to empty string? What does it return in that case?)

File: soundcard/pulseaudio.py (hunk 1 above)
Comment (flying-sheep) [reply soundcard/pulseaudio.py:93]: Seems to be not the case, this is probably a preprocessor macro instead …

File: soundcard/pulseaudio.py (hunk 2 above)
Comment (flying-sheep) [reply soundcard/pulseaudio.py:101]: > Does pulse guarantee/enforce a valid UTF-8 name?

yes.

> Can you set the name to empty string? What does it return in that case?

yes, and it returns an empty string.

**Comment by flying-sheep** [comment]
I went for `os.fsencode(os.path.basename(sys.argv[0]))`, because command lines can get pretty long. But of course I’m open to go for preserving the options if you want.

The only problem with my approach I can think of is when someone uses `python -c 'import soundcard; ...'`, but I think that’s an edge case, don’t you agree?

**Comment by bastibe** [comment]
I agree. As long as it's reasonably obvious to the user that this is the Python script they're running, a short name is better than a longer one.

Could you give an example of how your function would handle, say `python mycoolscript.py` and `./mycoolscript.py`?

As this functionality is currently pulseaudio/Linux-only, I'd prefer to keep it inside pulseaudio.py instead of a separate file for now.

**Comment by flying-sheep** [comment]
```console
$ printf "#\!/usr/bin/env python3\nimport sys, os\nprint(os.path.basename(sys.argv[0]))" >! /tmp/mycoolscript.py
$ cd /tmp
$ chmod +x mycoolscript.py
$ python mycoolscript.py
mycoolscript.py
$ ./mycoolscript.py
mycoolscript.py
```

**Comment by bastibe** [comment]
Cool, thank you very much!

This looks very good to me! Only one small change is still needed, that I missed last time (sorry): The documentation for SoundCard is built from the Linux source, so the docstrings of `set_name` and `get_name` need to mention that they are only implemented for Linux at the moment.

**Comment by flying-sheep** [comment]
Done. I also edited the doc config:

https://github.com/flying-sheep/SoundCard/blob/c39b2bb80f4cf59e338f14a19eef5f360d223781/doc/conf.py#L184-L191

It didn’t work for me, as it infinitely hung here:

https://github.com/flying-sheep/SoundCard/blob/c39b2bb80f4cf59e338f14a19eef5f360d223781/soundcard/pulseaudio.py#L70

**Comment by bastibe** [comment]
Great work! Thank you.

Do you feel this is ready to merge?

**Comment by flying-sheep** [comment]
Yeah, I think so!

**Comment by bastibe** [comment]
Thank you very much for your contribution, and a pleasant conversation!

(Should I forget to cut a new release for SoundCard in a month or so, feel free to ping me on this thread.)

**Pull request merged**